photon-lib/src/main/scala/com/linkedin/photon/ml/function/PriorDistribution.scala

```diff
@@ -0,0 +1,256 @@
+ * regularization weight refers to the relative weight of the Gaussian prior.
+ */
+trait PriorDistribution extends ObjectiveFunction {
+
+  val priorCoefficients: ModelCoefficients = ModelCoefficients(DenseVector.zeros(1))
+
+  lazy protected val priorMeans: Vector[Double] = priorCoefficients.means
+  lazy protected val priorVariances: Vector[Double] = priorCoefficients.variancesOption.get
+  lazy protected val inversePriorVariances: DenseVector[Double] = priorVariances.map(v => if (v > MathConst.EPSILON) 1.0 / v else 1.0).toDenseVector
+  protected var l1RegWeight: Double = 0D
+  protected var l2RegWeight: Double = 0D
+
+  require(l1RegWeight >= 0D, s"Invalid regularization weight '$l1RegWeight")
+  require(l2RegWeight >= 0D, s"Invalid regularization weight '$l2RegWeight")
+
+  /**
+   * Getter for the Laplace weight of the prior.
+   *
+   * @return The L1 regularization weight
+   */
+  def l1RegularizationWeight: Double = l1RegWeight
+
+  /**
+   * Getter for the Gaussian weight of the prior.
+   *
+   * @return The L2 regularization weight
+   */
+  def l2RegularizationWeight: Double = l2RegWeight
+
+  /**
+   * Compute the value of the function over the given data for the given model coefficients, with regularization towards
+   * the prior coefficients.
+   *
+   * @param input The data over which to compute the objective function value
+   * @param coefficients The model coefficients for which to compute the objective function's value
+   * @param normalizationContext The normalization context
+   * @return The value of the objective function and regularization terms
+   */
+  abstract override protected[ml] def value(
+      input: Data,
+      coefficients: Coefficients,
+      normalizationContext: BroadcastWrapper[NormalizationContext]): Double =
+    super.value(input, coefficients, normalizationContext) +
+      l1RegValue(convertToVector(coefficients)) +
+      l2RegValue(convertToVector(coefficients))
+
+  /**
+   * Compute the Laplace regularization term for the given model coefficients. L1 regularization term is
+   * l1RegWeight * sum(abs(coefficients - priorMeans) :/ sqrt(priorVariance)).
+   *
+   * @param coefficients The model coefficients
+   * @return The Laplace regularization term value
+   */
+  protected def l1RegValue(coefficients: Vector[Double]): Double = {
+
+    val normalizedCoefficients = (coefficients - priorMeans) *:* sqrt(inversePriorVariances)
+
+    l1RegWeight * sum(abs(normalizedCoefficients))
+  }
+
+  /**
+   * Compute the Gaussian regularization term for the given model coefficients. L2 regularization term is
+   * l2RegWeight * sum(pow(coefficients - priorMeans, 2) :/ priorVariance) / 2.
+   *
+   * @param coefficients The model coefficients
+   * @return The Gaussian regularization term value
+   */
+  protected def l2RegValue(coefficients: Vector[Double]): Double = {
+
+    val normalizedCoefficients = (coefficients - priorMeans) *:* sqrt(inversePriorVariances)
+
```

Comment:
why sqrt, then square later? why not multiple it  at line 98

Reply:
These are equivalent. We can merge the 2 lines together

Reply:
fixed